### DISCUSSION TITLE
[Question]: Coral edgeTPU: why has inference "speed" has improved with use of an F+ model ?

### DISCUSSION THREAD

**Discussion by ElectronicBattle** [blakeblackshear/frigate#13419] [Ask A Question] [answered]
### What is your question?

Hello all

Why does the use of an F+ model improve inference processing: how does it allow the inference times to reduce i.e. the inference "speed" increases?

My USB Coral was typically showing inference speeds of about 9-10.5 milliseconds on my untrained system. 

Yesterday, having installed my first proper model (and having made no other changes) I noted that the time had reduced to approx 7-8 ms. 

Today, after I'd added some filtering and restarting Frigate, I'm seeing further improvement with a reduction in time down to perhaps 6-7 ms.  

This is something like a 33% improvement, in other words. Is it the case that the decision boundaries or thresholds are 

(i) now higher due to the F+ model being present,  and then 
(ii) I have also added filtering, and hence fewer detections are being selected for Coral assessment?

**Comment by NickM-27** [comment]
The inference speed is faster because the model is more efficient

**Reply by ElectronicBattle** [reply, marked as answer]
I read up a bit more (AI engine/Google etc.) about this sort of general area of machine learning, and I've summarised it for other users because I think it might be useful to see a broad answer. If what I have written is generally correct it'll hopefully be useful but if it's wrong let me know and I'll delete it (or you can delete it as you see fit)

The Frigate+ model might be optimised for my Coral edgeTPU because:

(i) The improved accuracy of the Frigate+ model might generate fewer false positives or fewer classification errors, reducing the overall number of bounding boxes and object data that needs to be processed, which  contributes to faster inference speed (effectively some data reduction due to better accuracy == "more efficient").

(ii) The model's performance is not solely dependent on the number of training images for each individual CCTV camera. Rather, the model might be learning general features and patterns from the sum training data I have provided (all 1080+ images), which can be then applied across multiple cameras. It learns a comprehensive understanding of objects and their variations based on the set of all images, not just per camera. This learned knowledge can then be used to detect and classify objects in new, unseen images, such as those from cameras with fewer training images.  Those cameras with fewer images submitted for them ("fewer" being a couple of dozen) benefit from the processing of camera images where hundreds have been submitted.

The F+ model might employ "transfer learning" which is a technique whereby a pre-trained model is fine-tuned with a smaller dataset. This allows the model to apply initial training data convolved with  newer tasks which themselves contain  fewer training examples in some cases (i.e. cameras with not many associated images vs cameras with lots of associated images).
